activerecord/lib/active_record/relation.rb

```diff
@@ -304,84 +304,151 @@ def many?
+            last_id_offset = connection.adapter_name == "PostgreSQL" ?
+              Arel::Nodes::NamedFunction.new("GREATEST", [collection_size_minus_one, 0]) :
+              aggregates_table.project(collection_size_minus_one)
+
+            select_values = [
+              aggregates_table[:collection_size],
+              aggregates_table[:timestamp],
+              collection_table.project(collection_table[:collection_cache_key_id]).take(1).as("first_id"),
+              collection_table.project(collection_table[:collection_cache_key_id]).take(1).skip(last_id_offset).as("last_id")
+            ]
+
+            arel = aggregates_table
+              .project(select_values)
+              .with(collection_common_table_expr, aggregates_common_table_expr)
+          when "Mysql2"
+            select_values = "COUNT(*) AS collection_size, @first_id AS first_id, @last_id AS last_id, MAX(collection_cache_key_timestamp) AS timestamp"
+            query = collection.select("COALESCE(@first_id, @first_id := #{id_col}), @last_id := #{id_col}, #{timestamp_col} AS collection_cache_key_timestamp")
+            subquery_alias = "subquery_for_cache_key"
+            arel = query.build_subquery(subquery_alias, select_values)
+          else
+            select_values = "COUNT(*) AS collection_size, MAX(#{timestamp_col}) AS timestamp"
+            query = collection.select("#{timestamp_col} AS collection_cache_key_timestamp")
+            subquery_alias = "subquery_for_cache_key"
+            arel = query.build_subquery(subquery_alias, select_values)
+          end
         else
+          select_values = "COUNT(*) AS collection_size, MAX(#{timestamp_col}) AS timestamp"
           query = collection.unscope(:order)
-          query.select_values = [select_values % column]
+          query.select_values = [select_values]
           arel = query.arel
         end
 
-        result = connection.select_one(arel, nil)
+        result = nil
+
+        connection_pool.with_connection do
+          if connection.adapter_name == "Mysql2" && collection.has_limit_or_offset?
+            connection.execute("SET @first_id := @last_id := NULL")
+          end
+
+          result = connection.select_one(arel, nil)
+        end
 
         if result
           column_type = klass.type_for_attribute(timestamp_column)
-          timestamp = column_type.deserialize(result["timestamp"])
-          size = result["size"]
-        else
-          timestamp = nil
-          size = 0
+          size = result["collection_size"]
+
+          if size > 0
+            first_id = result["first_id"]
+            last_id = result["last_id"]
+            timestamp = column_type.deserialize(result["timestamp"]).utc.to_s(cache_timestamp_format)
+          end
         end
       end
 
-      if timestamp
-        "#{size}-#{timestamp.utc.to_s(cache_timestamp_format)}"
-      else
-        "#{size}"
-      end
+      [size, first_id, last_id, timestamp].reject(&:blank?).join("-")
```

Comment:
Is there any harm in leaving blank values in the key, i.e. `"#{size}-#{first_id}-#{last_id}-#{timestamp}"`?  That would save two array allocations.

Reply:
I'd like to avoid making necessary changes to the existing cache key algorithm, per DHH's comment, [here](https://github.com/rails/rails/pull/29092#issuecomment-301503121):

> Furthermore, any change to CacheStore has to be backwards compatible. If
you supply the same values and get a different key, that's no bueno.

But, here's a slightly less succinct solution that avoids necessary array allocations:

```ruby
case
when size.zero?
  "0"
when first_id
  "#{size}-#{first_id}-#{last_id}-#{timestamp}"
else
  "#{size}-#{timestamp}"
end
```

I've rebased accordingly, thanks for flagging.